Steady-state Lotus Molten Chloride Reactor - Refs #313 (#314)
idaholab/virtual_test_bed: doc/content/msr/lotus/lotus_multiphysics_model.md

# Lotus Griffin-Pronghorn Model

*Contact: Mauricio Tano, mauricio.tanoretamales.at.inl.gov*

*Model summarized, documented, and uploaded by Rodrigo de Oliveira and Samuel Walker*

*Model link: [Griffin-Pronghorn Steady-State Lotus Model](https://github.com/idaholab/virtual_test_bed/tree/devel/msr/lotus/steady_state)*

!tag name=Lotus Griffin-Pronghorn Steady State Model pairs=reactor_type:MSR

>                        reactor:Lotus

Reviewer: ```suggestion
                       reactor:generic-msr
```

!row!
!media msr/lotus/MCR_geometry.jpg
       style=width:50%;float:left;padding-top:2.5%;padding-right:5%
       id=LMCR_pgh_geometry
       caption=LMCR geometry for neutronics evaluation [!citep](M3mcr2023).

!media msr/lotus/mcr_thermal_hydraulics.png
       style=width:45%;float:left;padding-top:2.5%;padding-right:5%
       id=LMCR_pgh_thermal_hydraulics
       caption=Thermal-hydraulics blocks for LMCR thermal-hydraulics & DNP domain [!citep](M3mcr2023).
!row-end!

## Neutronics

Starting first with neutronics, Griffin solves the neutron transport problem via the Diffusion equation approximation. The Griffin input file will now be briefly discussed and the primary equations that are solved and how they relate to the input file will be shown. The input file to solve the 9 group neutron diffusion problem is shown below.

> This is the main input file which calls the open source MOOSE Navier-Stokes input file `run_ns_initial_res.i` which is the sub-app for solving the Thermal Hydraulics solution of this model.

Reviewer: ```suggestion
This is the main input file which calls the open source MOOSE Navier-Stokes input file `run_ns_initial.i` which is the sub-app for solving the Thermal Hydraulics solution of this model.
```

!listing msr/lotus/steady_state/run_neutronics_9_group.i

#### Mesh

> Starting first with the mesh, this block defines the computational domain that the neutronics solve will operate on. Here, a cubit generated 3D mesh is imported as an exodus file. It is correspondingly manupiulated to add in a reactor boundary condition, and is scaled to the correct size of the reactor.

Reviewer: ```suggestion
Starting first with the mesh, this block defines the computational domain that the neutronics solve will operate on. Here, a cubit generated 3D mesh is imported as an exodus file. It is correspondingly manipulated to add in a reactor boundary condition, and is scaled to the correct size of the reactor.
```

#### Power Density

Next, the `Power Density` is set for this specific reactor. Here the neutronics solution is normalized to the rated power of $25$kW.

!listing msr/lotus/steady_state/run_neutronics_9_group.i block=PowerDensity

The power density is evaluated with the normalized neutronics solution.
It provides the energy source in the fluid energy conservation equation.
Because the fluid energy equation is discretized with FV, we evaluate the power
density variable with a constant monomial function.

#### Auxiliary Variables

Moving on, the `AuxVariables` are now set which are specific variables that can be passed and read in from the other sub-apps for multiphysics calculations.

!listing msr/lotus/steady_state/run_neutronics_9_group.i block=AuxVariables

Here the `tfuel` and `trefl` are the temperature of the fuel salt and reflector, whereas `c1` - `c6` are the DNP distributions, and `dnp` is an array auxiliary with 6 component corresponding
to the 6 DNP groups used here. This array is then read into Griffin in the `external_dnp_variable = 'dnp'` parameter within the `Transport Systems` block.

#### Auxiliary Kernels

> Correspondingly, the `AuxKernels` are functors which operate on the `AuxVariables`. Here the six seperate DNP groups are compiled into the Aux Variable `dnp` to be used within Griffin.

Reviewer: ```suggestion
Correspondingly, the `AuxKernels` are functors which operate on the `AuxVariables`. Here the six separate DNP groups are compiled into the Aux Variable `dnp` to be used within Griffin.
```

`CoupledFeedbackNeutronicsMaterial` is able to use the temperature transferred
from the fluid system for evaluating multigroup cross sections based on a table lookup
on element quadrature points to bring in the feedback effect.

#### Executioner

Next, the `Executioner` block sets up the type of problem, and the numerical methods to solve the neutronics and multiphysics problems.

!listing msr/lotus/steady_state/run_neutronics_9_group.i block=Executioner

The caculation is driven by `Eigenvalue`, an executioner available in Griffin.
The PJFNKMO option for the `solve_type` parameter is able to
drive the eigenvalue calculation with the contribution of DNP
to the neutron transport equation as an external source scaled with $k$-effective.

Additionaly, PETSc options and tolerances for the neutronic and multiphysics fixed point iteration coupling method are provided.

#### Post Processors

The `PostProcessors` block sets up various calculations of reactor parameters that may be of interest to the user. This can be helpful to ensure the model is implemented correctly. Here the average, maximum, and minimum of various variables (e.g., `power`, `fluxes`, and `DNPs`) can be computed.

> !listing msr/lotus/steady_state/run_neutronics_9_group.i block=Postprocessors

Reviewer: ```suggestion
!listing msr/lotus/steady_state/run_neutronics_9_group.i block=Postprocessors VectorPostprocessors
```

input file shown below. (Here "ns" is an abbreviation for Navier-Stokes.)

The fluid system includes conservation equations for fluid mass, momentum, and
energy. Here a porous flow and weakly-compressible formulation are used to model molten salts and the pressure drop over the mixing plate at the entrance of the reactor core.

Additionally, the thermal hydraulics input file has another sub-app below it that calculates the delayed neutron precursor (DNP) group distributions in `run_prec_transport.i` which will be discussed shortly.

!listing msr/lotus/steady_state/run_ns_initial_res.i

#### Problem Parameters

The physical properties of the fuel salt and reflector (e.g., `density`, `viscosity`, `thermal conductivity`, and `heat capacity`) are defined first. Additionally, other parameters such as a friction force, pump force, porosity, and area are also set. Lastly, numerical interpolation schemes such as `upwind` and `rhie-chow` are also selected and a mixing length turbulence calibration informed from high-fidelity CFD is also set here.

!listing msr/lotus/steady_state/run_ns_initial_res.i start=# Material Properties end=[GlobalParams]

#### Global Parameters

> Next, `Global Parameters` lists variables and parameters that will be used throughout the entire input file. Here the finite volume interpolation methods are set, and `superficial velocities`, `pressure`, `porosity`, `density`, `viscosity`, and the `mixing_length` model are defined for the entire input file. Some of these values are originally set in the `Problem Parameters` and are referenced here using the ${} notation to be used implicitly for the rest of the input file.

Reviewer: ```suggestion
Next, `Global Parameters` lists variables and parameters that will be used throughout the entire input file. Here the finite volume interpolation methods are set, and `superficial velocities`, `pressure`, `porosity`, `density`, `viscosity`, and the `mixing_length` model are defined for the entire input file. Some of these values are originally set in the `Problem Parameters` / header of the input and are referenced here using the ${} notation to be used implicitly for the rest of the input file.
```

!listing msr/lotus/steady_state/run_ns_initial_res.i block=AuxVariables

#### Auxiliary Kernels

Correspondingly, the `Auxiliary Kernels` operate on the `Auxiliary Variables`. In this case the auxiliary variables needed by the precursor advection sub-app are generated by copying the superficial velocities given by the thermal hydraulics solution.

!listing msr/lotus/steady_state/run_ns_initial_res.i block=AuxKernels

#### Functions

Next, various functions are defined to be used throughout the input file. Here heat transfer coefficients, reynolds numbers, power-density functions, and cosine guesses can all be generated and used for initial conditions, kernels, etc.

!listing msr/lotus/steady_state/run_ns_initial_res.i block=Functions

#### Materials

Next, the `Materials` block specifies material properties of interest in the fuel salt, reactor vessel, and reflector. Here, `viscosity`, `wall frictions`, and turbulent `mixing length` models can be implemented.

!listing msr/lotus/steady_state/run_ns_initial_res.i block=Materials

#### Executioner

> The `Executioner` block defines how the problem will be run. Here the type of solution is a `Transient` and a `TimeStepper` function is defined. Additionally, the numerical tolerances for the solution are selected to speed up convergence.

Reviewer: reword

> The `Executioner` block defines how the problem will be run. Here the type of solution is a `Transient` and a `TimeStepper` function is defined. Additionally, the numerical tolerances for the solution are selected to speed up convergence.

Reviewer: ```suggestion
The `Executioner` block defines how the problem will be run. Here the type of solution is a `Transient` and a `TimeStepper` function is defined. Additionally, the numerical tolerances for the solution are selected to control the level of convergence required.
```
